### DISCUSSION TITLE
自定义sendMessageHook被执行多次？

### DISCUSSION THREAD

**Discussion by dousp** [apache/rocketmq#7518] [Q&A]
## 描述
- org.apache.rocketmq:rocketmq-spring-boot-starter:2.2.3
- 注册了了一个自定义的`SendMessageHook`
- 发送异步消息的时候，发现`sendMessageAfter`被调用了两次
![image](https://github.com/apache/rocketmq/assets/5936499/b1423204-6f8e-4ee9-83d4-a50096d51cbc)

- 我查了下源码，发现两个地方都有调用，并且发送异常时可能也会导致`sendMessageAfter`重复执行两遍
- 这里应该是有问题的吧？

![image](https://github.com/apache/rocketmq/assets/5936499/9d566ef3-2817-4f2f-8037-c3127ceb127f)
![image](https://github.com/apache/rocketmq/assets/5936499/887268be-bc07-4645-b816-c2f117ed8e54)
![image](https://github.com/apache/rocketmq/assets/5936499/7dd13435-c38a-4ec8-88ab-854db29bcc6e)
![image](https://github.com/apache/rocketmq/assets/5936499/4aa2dbab-1ccc-4f21-b5c8-bf8cbb230183)

**Comment by figuewang** [comment]
是的因该把默认的hook替换掉 ，没看见是怎么初始化的

**Comment by leizhiyuan** [comment]
使用了异步发送？异步发送的时候，hook确实会执行两次，异步里面一次，外面一次，但是外面因为sendResult为空，所以实际上执行了进去会直接跳出。

**Reply by dousp** [reply]
应该是异步发送，回头我再调试看下。
